### ISSUE TITLE
Passenger 6.0.27 – spawn failure with errno=24, very high number of open FDs

### ISSUE THREAD

**Issue by Munrok** [phusion/passenger#2637]
We’re running Passenger on a shared hosting environment, where users run their own applications via Passenger. On Passenger 6.0.27 we’re seeing cases where an application fails to start.
From passenger.log:
```
[ E 2025-12-14 12:58:08.0616 58412/Tjw3 age/Cor/App/Implementation.cpp:218 ]:
Could not spawn process for application <REDACTED_APP_PATH>:
An operating system error occurred while preparing to spawn an application process:
Error looking up OS group account 1000: Too many open files (errno=24)
```

At the same time, PassengerAgent ends up with a very large number of open file descriptors (around 80k+).
System limits are high (openfiles 3771288), so it doesn’t look like a ulimit problem.

On Passenger 6.0.22 this setup was working fine: PassengerAgent typically had around 200 open files, not tens of thousands.

After attaching truss to PassengerAgent, the process crashed, and after restart the problem temporarily disappeared.

**Comment by CamJN** [comment]
Between `release-6.0.22` and `release-6.0.27` these files had changes that involve fds:

- `src/agent/Core/ApplicationPool/Socket.h` now `connection.fd = state.getFd().detach();` is set differently, was `connection.fd = connectToServer(address, __FILE__, __LINE__); `
- `src/agent/Core/ApplicationPool/Session.h` `virtual bool initiate() override` no longer sets the fd as non-blocking and therefore doesn't have a fdguard
- `src/agent/SpawnEnvSetupper/SpawnEnvSetupperMain.cpp` changed from boost bind `boost::bind(reopenStdout, fileno(f))); ` to closure `int fd = fileno(f);    runCommand(command, info, true, true, [=]() { reopenStdout(fd);  });`
- `src/agent/Core/Controller/CheckoutSession.cpp` `onSessionSocketConnected` and `initiateSession` have new uses of fds
- `src/cxx_supportlib/IOTools/IOUtils.*` `NConnect_State::getFd()` is used to get fd instead of `connectToServer(NConnect_State &state)` interestingly `connectToServer` used to return a `bool`...

**Comment by CamJN** [comment]
You can also set `PassengerFileDescriptorLogFile` to log fd activity, which might help gather info.

**Comment by Munrok** [comment]
@CamJN In the additional log, there is no information about the buffer files.

**Comment by CamJN** [comment]
The fds might not be buffer-related, really we're just looking for fd leaks, that is to say fds that are opened but not closed.

**Comment by Munrok** [comment]
@CamJN I’d like to add that the files are never deleted and they remain on disk. Currently, on one of the servers there are 212,027 buffer files. Restarting Passenger doesn’t remove them either.

**Comment by CamJN** [comment]
Okay that's new info, what's the path for these buffer files?

**Comment by Munrok** [comment]
@CamJN I’m forwarding an excerpt from the FDS logs:

[ D 2025-12-17 13:12:02.9511 38065/T3bn FileDescriptor.h:84 ]: File descriptor closed: 5273
[ D 2025-12-17 13:12:03.0512 38065/T3bn FileDescriptor.h:84 ]: File descriptor closed: 854
[ D 2025-12-17 13:12:03.0514 38065/T3bo FileDescriptor.h:94 ]: File descriptor closed: 1088
[ D 2025-12-17 13:12:03.1515 38065/T3bo FileDescriptor.h:84 ]: File descriptor closed: 1089
[ D 2025-12-17 13:12:03.1517 38065/T3bp FileDescriptor.h:94 ]: File descriptor closed: 410
[ D 2025-12-17 13:12:03.3518 38065/T3bp age/Cor/App/Socket.h:69 ]: File descriptor closed: 183
[ D 2025-12-17 13:12:03.3519 38065/T3bp FileDescriptor.h:84 ]: File descriptor closed: 411
[ D 2025-12-17 13:12:03.3521 38065/T3bq FileDescriptor.h:94 ]: File descriptor closed: 951
[ D 2025-12-17 13:12:03.4522 38065/T3bq FileDescriptor.h:84 ]: File descriptor closed: 952
[ D 2025-12-17 13:12:03.4524 38065/T3br FileDescriptor.h:94 ]: File descriptor closed: 466
[ D 2025-12-17 13:12:03.5525 38065/T3br FileDescriptor.h:84 ]: File descriptor closed: 467
[ D 2025-12-17 13:12:03.5527 38065/T3bs FileDescriptor.h:94 ]: File descriptor closed: 759
[ D 2025-12-17 13:12:03.5527 38065/T3bs FileDescriptor.h:94 ]: File descriptor closed: 2338
[ D 2025-12-17 13:12:03.6528 38065/T3bs FileDescriptor.h:84 ]: File descriptor closed: 760
[ D 2025-12-17 13:12:03.6529 38065/T3bs FileDescriptor.h:84 ]: File descriptor closed: 2339
[ D 2025-12-17 13:12:03.6532 38065/T3bu FileDescriptor.h:94 ]: File descriptor closed: 3123
[ D 2025-12-17 13:12:03.7533 38065/T3bu FileDescriptor.h:84 ]: File descriptor closed: 3125
[ D 2025-12-17 13:12:03.7535 38065/T3bv FileDescriptor.h:94 ]: File descriptor closed: 1002
[ D 2025-12-17 13:12:03.8536 38065/T3bv FileDescriptor.h:84 ]: File descriptor closed: 1005
[ D 2025-12-17 13:12:03.8539 38065/T3bx FileDescriptor.h:94 ]: File descriptor closed: 2815

Replying to the second message: the path is /tmp/buffer.[0-9]+$, for example:

24752 PassengerAgent    8479 v r rw------   1       0 -   /tmp/buffer.264401959
24752 PassengerAgent    8488 v r rw------   1       0 -   /tmp/buffer.1740717856
24752 PassengerAgent    8489 v r rw------   1       0 -   /tmp/buffer.624240903
24752 PassengerAgent    8490 v r rw------   1       0 -   /tmp/buffer.1258359673
24752 PassengerAgent    8491 v r rw------   1       0 -   /tmp/buffer.391463133
24752 PassengerAgent    8494 v r rw------   1       0 -   /tmp/buffer.1030551195
24752 PassengerAgent    8497 v r rw------   1       0 -   /tmp/buffer.1360524207
24752 PassengerAgent    8499 v r rw------   1       0 -   /tmp/buffer.1850593327
24752 PassengerAgent    8504 v r rw------   1       0 -   /tmp/buffer.794816567
24752 PassengerAgent    8505 v r rw------   1       0 -   /tmp/buffer.1905793042
24752 PassengerAgent    8511 v r rw------   1       0 -   /tmp/buffer.646542773
24752 PassengerAgent    8512 v r rw------   1       0 -   /tmp/buffer.821583572
24752 PassengerAgent    8513 v r rw------   1       0 -   /tmp/buffer.995117768
24752 PassengerAgent    8529 v r rw------   1       0 -   /tmp/buffer.1547860596
24752 PassengerAgent    8551 v r rw------   1       0 -   /tmp/buffer.982383933
24752 PassengerAgent    8599 v r rw------   1       0 -   /tmp/buffer.1861193336
24752 PassengerAgent    8622 v r rw------   1       0 -   /tmp/buffer.357228111
24752 PassengerAgent    8625 v r rw------   1       0 -   /tmp/buffer.1674120996

**Comment by samupl** [comment]
Looks like a similar issue was already reported as #2568

**Comment by CamJN** [comment]
If we assume that the issue is the same, and it does look very similar, then that bounds the problem to a change introduced in version 6.0.23 since @Munrok said it doesn't occur in 6.0.22. However that doesn't seem to line up with the actual changes in Passenger.

By running `git diff --name-only  release-6.0.2{2,3} -- src | rg -vFe AutoGenerated -e '.rb' | xargs git diff release-6.0.2{2,3} --`
I can see that the only non-comment c++ changes are the addition of a custom error page renderer, which doesn't involve a FileBufferedChannel.

**Comment by CamJN** [comment]
As for the fds logs, it's not really helpful to just get an excerpt like that, what I need is to be able to trace the fds that are opened and not closed. So what I usually do is for each fd closed line like you copied into the excerpt, I delete all the logs that refer to that fd. Then I'm left with the logs pertaining to fds that were not closed.

**Comment by Munrok** [comment]
@CamJN I tested different setups to figure out which configuration triggers the issue.

The problem does not appear with the following setups:
- nginx 1.28.0 + passenger 6.0.22 + libuv 1.48.0
- nginx 1.28.0 + passenger 6.0.27 + libuv 1.48.0
- nginx 1.28.0 + passenger 6.0.27 + libuv 1.49.2

However, the bug does occur with these setups (bug: file descriptor is not being closed):
- nginx 1.28.0 + passenger 6.0.22 + libuv 1.51.0
- nginx 1.28.0 + passenger 6.0.27 + libuv 1.51.0
- nginx 1.28.0 + passenger 6.0.27 + libuv 1.50.0

**Comment by CamJN** [comment]
Ah well that explains a lot, we do not support external copies of libuv. We vendor the library so that we use a known working copy. If you change that it's not surprising that it breaks things.

**Comment by Munrok** [comment]
@CamJN Thank you for your help, they were able to fix it in the FreeBSD ports thanks to you.
